src/main/java/sirius/biz/process/ProgressTracker.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Made with all the love in the world
+ * by scireum in Remshalden, Germany
+ *
+ * Copyright by scireum GmbH
+ * http://www.scireum.de - info@scireum.de
+ */
+
+package sirius.biz.process;
+
+import sirius.kernel.nls.NLS;
+
+import javax.annotation.Nullable;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.time.temporal.ChronoUnit;
+import java.util.Optional;
+
+/**
+ * Defines a class used to track the progress of the migration and calculates the estimated completion time.
+ */
+public class ProgressTracker {
+    private final long total;
+    private final LocalDateTime start;
+    private long current = 0;
+
+    private boolean finished = false;
+
+    private String message = null;
+
+    /**
+     * Initializes the progress tracker with a total amount of units.
+     *
+     * @param total the total amount of units to track
+     * @return the progress tracker
+     */
+    public static ProgressTracker start(long total) {
+        return new ProgressTracker(total);
+    }
+
+    private ProgressTracker(long total) {
+        if (total < 0) {
+            throw new IllegalArgumentException("Total must be >= 0");
+        }
+        this.start = LocalDateTime.now();
+        this.total = total;
+    }
+
+    /**
+     * Increments the processed unit by one and recalculates the progress.
+     *
+     * @param process the process context to update the state message
+     */
+    public void increment(ProcessContext process) {
+        increment(process, 1);
+    }
```

Comment:
**Optional:** Could add a `withProcessContext` method which sets the context in the tracker and then `assertProcessContext` to check if it was invoked (which happens automatically when calling `startTracking`)